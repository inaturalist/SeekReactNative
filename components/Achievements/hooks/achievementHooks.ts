// @flow

import { useState, useEffect } from "react";
import Realm from "realm";
import { addMonths, isEqual } from "date-fns";

import realmConfig from "../../../models";
import { taxonIds } from "../../../utility/dictionaries/taxonomyDicts";

const useFetchAchievements = ( ): any => {
  const [state, setState] = useState( {
    speciesBadges: [],
    level: null,
    nextLevelCount: 0,
    badgesEarned: null
  } );

  useEffect( ( ) => {
    const fetchAchievements = async ( ) => {
      try {
        const realm = await Realm.open( realmConfig );
        const badges = realm.objects( "BadgeRealm" );
        const badgesEarned = badges.filtered( "iconicTaxonName != null AND earned == true" ).length;
        const iconicTaxonIds = Object.keys( taxonIds ).map( id => taxonIds[id] );

        const speciesBadges = [];

        iconicTaxonIds.forEach( ( id ) => {
          if ( id === null ) { return; }
          const highestEarned = badges
            .filtered( `iconicTaxonName != null AND iconicTaxonId == ${id}` )
            .sorted( "index", true )
            .sorted( "earned", true );
          speciesBadges.push( highestEarned[0] );
        } );

        const levelsEarned = badges.filtered( "iconicTaxonName == null AND earned == true" ).sorted( "count", true );
        const nextLevel = badges.filtered( "iconicTaxonName == null AND earned == false" ).sorted( "index" );

        setState( {
          speciesBadges,
          level: levelsEarned[0],
          nextLevelCount: nextLevel[0] ? nextLevel[0].count : 0,
          badgesEarned
        } );
      } catch ( e ) {
        console.log( e, "couldn't open realm: achievements" );
      }
    };
    fetchAchievements( );
  }, [] );

  return state;
};

const useFetchChallenges = ( ): any => {
  const [challengeBadges, setChallengeBadges] = useState( [] );

  useEffect( ( ) => {
    const createBadge = ( latestBadge, numOfMonths ) => ( {
      name: "",
      availableDate: addMonths( latestBadge.availableDate, numOfMonths ),
      index: latestBadge.index + numOfMonths
    } );

    const createPlaceholderBadges = ( badges ) => {
      const remainderOfBadges = badges.length % 5;

      if ( remainderOfBadges === 0 || remainderOfBadges === 3 ) {
        // no placeholders needed
        return badges;
      }

      const badgePlaceholders = badges;
      const latestBadge = badges[badges.length - 1];

      let nextBadge = createBadge( latestBadge, 1 );

      // next challenge after June 2021 will be released Aug 2021
      if ( isEqual( new Date( 2021, 5, 1 ), new Date( latestBadge.availableDate ) ) ) {
        nextBadge = createBadge( latestBadge, 2 );
      }

      const badgeAfterNext = createBadge( nextBadge, 1 );

      if ( remainderOfBadges === 2 || remainderOfBadges === 4 ) {
        badgePlaceholders.push( nextBadge );
      }

      if ( remainderOfBadges === 1 ) {
        badgePlaceholders.push( nextBadge, badgeAfterNext );
      }

      return badgePlaceholders;
    };

    const fetchChallenges = async ( ) => {
      try {
        const realm = await Realm.open( realmConfig );
        const challenges = realm.objects( "ChallengeRealm" ).sorted( "availableDate", false );
        const badges = challenges.map( ( challenge ) => challenge );
        const badgesWithPlaceholders = createPlaceholderBadges( badges );
        setChallengeBadges( badgesWithPlaceholders );
      } catch ( e ) {
        console.log( e, "couldn't open realm: achievements" );
      }
    };
    fetchChallenges( );
  }, [] );

  return challengeBadges;
};

export {
  useFetchAchievements,
  useFetchChallenges
};
