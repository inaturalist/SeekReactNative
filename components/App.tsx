import React, { useEffect } from "react";
import {
  setJSExceptionHandler,
  setNativeExceptionHandler
} from "react-native-exception-handler";
import {
  getVersion,
  getBuildNumber
} from "react-native-device-info";
import { SafeAreaProvider } from "react-native-safe-area-context";
import { GestureHandlerRootView } from "react-native-gesture-handler";

import RootStack from "./Navigation/RootStack";
import { hideLogs } from "../utility/helpers";
import { LanguageProvider } from "./Providers/LanguageProvider";
import { SpeciesNearbyProvider } from "./Providers/SpeciesNearbyProvider";
import UserLoginProvider from "./Providers/UserLoginProvider";
import { ObservationProvider } from "./Providers/ObservationProvider";
import { AppOrientationProvider } from "./Providers/AppOrientationProvider";
import { ChallengeProvider } from "./Providers/ChallengeProvider";
import { SpeciesDetailProvider } from "./Providers/SpeciesDetailProvider";
import { log } from "../react-native-logs.config";
import { LogLevels, logToApi } from "../utility/apiCalls";
import ErrorBoundary from "./ErrorBoundary";

const logger = log.extend( "App.tsx" );

const jsErrorHandler = ( e, isFatal ) => {
  // not 100% sure why jsErrorHandler logs e.name and e.message as undefined sometimes,
  // but I believe it relates to this issue, which reports an unnecessary console.error
  // under the hood: https://github.com/a7ul/react-native-exception-handler/issues/143

  // possibly also related to error boundaries in React 16+:
  // https://github.com/a7ul/react-native-exception-handler/issues/60
  // if ( !e.name && !e.message ) {return;}
    if ( isFatal ) {
      logger.error( "Fatal JS Error: ", e.stack );
      logToApi( {
        level: LogLevels.ERROR,
        context: "App.tsx",
        message: e.message,
        errorType: e.constructor?.name,
        backtrace: e.stack
      } );
    } else {
      // This should get logged by ErrorBoundary. For some reason this handler
      // gets called too, so we don't want to double-report errors
    }

};

// record JS exceptions
setJSExceptionHandler( jsErrorHandler, true );

// record native exceptions
// only works in bundled mode; will show red screen in dev mode
// tested this by raising an exception in RNGestureHandler.m
// https://stackoverflow.com/questions/63270492/how-to-raise-native-error-in-react-native-app
setNativeExceptionHandler(
  async ( exceptionString ) => {
    try {
      logToApi( {
        level: LogLevels.ERROR,
        message: exceptionString,
        context: "App.tsx",
        errorType: "native"
      } );
      logger.error( `Native Error: ${exceptionString}` );
    } catch ( e ) {
      // Last-ditch attempt to log something
      logger.error(
        `Native Error: ${exceptionString} (failed to save context)`,
        e
      );
    }
  },
  true, // Force quit the app to prevent zombie states
  true // Enable on iOS
);

const style = { flex: 1 } as const;

const App = ( ) => {
  useEffect( () => {
    hideLogs( );
    logger.info( `App start. Version: ${getVersion()} Build: ${getBuildNumber()}` );
  }, [] );

  return (
    <AppOrientationProvider>
      <UserLoginProvider>
        <LanguageProvider>
          <ObservationProvider>
            <SpeciesNearbyProvider>
              <ChallengeProvider>
                <SpeciesDetailProvider>
                  <GestureHandlerRootView style={style}>
                    <SafeAreaProvider>
                      <ErrorBoundary>
                        <RootStack />
                      </ErrorBoundary>
                    </SafeAreaProvider>
                  </GestureHandlerRootView>
                </SpeciesDetailProvider>
              </ChallengeProvider>
            </SpeciesNearbyProvider>
          </ObservationProvider>
        </LanguageProvider>
      </UserLoginProvider>
    </AppOrientationProvider>
  );
};

export default App;
