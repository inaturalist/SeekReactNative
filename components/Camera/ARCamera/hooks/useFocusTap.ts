import {
  useCallback, useMemo, useRef, useState
} from "react";
import { Animated } from "react-native";
import {
  Gesture,
  GestureStateChangeEvent,
  TapGestureHandlerEventPayload
} from "react-native-gesture-handler";
import { Camera } from "../helpers/visionCameraWrapper";
const HALF_SIZE_FOCUS_BOX = 40;

interface Coordinates {
  x: number;
  y: number;
}

const useFocusTap = ( cameraRef: React.RefObject<Camera | null>, supportsFocus: boolean ) => {
  const [tappedCoordinates, setTappedCoordinates] = useState<Coordinates | null>( null );
  const focusOpacity = useRef( new Animated.Value( 0 ) ).current;

  const animatedStyle = useMemo( ( ) => {
    if ( !tappedCoordinates ) { return {}; }
    return ( {
      width: HALF_SIZE_FOCUS_BOX * 2,
      height: HALF_SIZE_FOCUS_BOX * 2,
      left: tappedCoordinates.x - HALF_SIZE_FOCUS_BOX,
      top: tappedCoordinates.y - HALF_SIZE_FOCUS_BOX,
      opacity: focusOpacity
    } );
  }, [tappedCoordinates, focusOpacity] );

  const onFocus = useCallback( async ( { x, y }: GestureStateChangeEvent<TapGestureHandlerEventPayload> ) => {
    // If the device doesn't support focus, we don't want the camera to focus
    if ( !supportsFocus ) {
      return;
    }
    cameraRef?.current?.focus( { x, y } );
    focusOpacity.setValue( 1 );
    setTappedCoordinates( { x, y } );
    Animated.timing(
      focusOpacity,
      {
        toValue: 0,
        duration: 2000,
        useNativeDriver: true
      }
    ).start( );
  }, [
    cameraRef,
    supportsFocus,
    focusOpacity
  ] );

  const tapToFocus = useMemo( ( ) => Gesture.Tap( )
    .runOnJS( true )
    .onStart( onFocus ), [onFocus] );

  return {
    animatedStyle,
    tapToFocus,
    tappedCoordinates
  };
};

export default useFocusTap;
