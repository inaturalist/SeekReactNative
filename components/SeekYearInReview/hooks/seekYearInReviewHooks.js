// @flow

import { useState, useEffect } from "react";
import Realm from "realm";
import { addMonths, isEqual } from "date-fns";

import realmConfig from "../../../models";

const useCountObservationsForYear = ( year ): any => {
  const [countObservationsThisYear, setCountObservationsThisYear] = useState( null );

  useEffect( () => {
    const fetchCount = async () => {
      try {
        const firstOfYear = () => new Date( year, 0, 1, 0, 0, 0, 0 );
        const lastOfYear = () => new Date( year, 11, 31, 23, 59, 59 );

        const realm = await Realm.open( realmConfig );
        const observations = realm.objects( "ObservationRealm" );
        const observationsThisYear = observations.filtered(
          "date >= $0 && date < $1",
          firstOfYear( year ),
          lastOfYear( year )
        );
        // TODO: Refactor this count out of this hook into seperate hook, because needed in home screen as well
        const count = observationsThisYear.length;
        setCountObservationsThisYear( count );
      } catch ( e ) {
        console.log( e, "couldn't open realm for counting observations for year" );
      }
    };
    fetchCount();
  }, [year] );

  return countObservationsThisYear;
};

const useObservationsForYear = ( year ): any => {
  const [observationsForYear, setObservationsThisYear] = useState( null );
  useEffect( () => {
    const fetchObservations = async () => {
      try {
        const firstOfYear = () => new Date( year, 0, 1, 0, 0, 0, 0 );
        const lastOfYear = () => new Date( year, 11, 31, 23, 59, 59 );

        const realm = await Realm.open( realmConfig );
        const observations = realm.objects( "ObservationRealm" );
        const observationsYear = observations.filtered(
          "date >= $0 && date < $1",
          firstOfYear( year ),
          lastOfYear( year )
        );
        setObservationsThisYear( observationsYear );
      } catch ( e ) {
        console.log( e, "couldn't open realm for fetching observations for year" );
      }
    };
    fetchObservations();
  }, [year] );

  return observationsForYear;
};

const useFetchStats = ( year ): any => {
  const [state, setState] = useState( {
    level: null,
    observationsThisYear: [],
    topThreeSpeciesBadges: [],
    randomObservations: [],
    histogram: [],
    // TODO: this is the same as in useFetchAchievements, refactor out
    speciesCount: null
  } );

  useEffect( () => {
    const fetchStatsFromRealm = async () => {
      try {
        const firstOfYear = () => new Date( year, 0, 1, 0, 0, 0, 0 );
        const lastOfYear = () => new Date( year, 11, 31, 23, 59, 59 );

        const realm = await Realm.open( realmConfig );

        const speciesCount = realm.objects( "TaxonRealm" ).length;

        // TODO: get observations from above useObservationsForYear hook
        const observations = realm.objects( "ObservationRealm" );
        const observationsThisYear = observations.filtered(
          "date >= $0 && date < $1",
          firstOfYear( year ),
          lastOfYear( year )
        );
        const countObservationsThisYear = observationsThisYear.length;

        function getRandom( arr, n ) {
          let len = arr.length;
          const result = new Array( n ),
            taken = new Array( len );
          if ( n > len )
            {throw new RangeError(
              "getRandom: more elements taken than available"
            );}
          while ( n-- ) {
            const x = Math.floor( Math.random() * len );
            result[n] = arr[x in taken ? taken[x] : x];
            taken[x] = --len in taken ? taken[len] : len;
          }
          return result;
        }
        // Get ten random observations from this year
        let randomObservations = [];
        if ( countObservationsThisYear > 10 ) {
          randomObservations = getRandom( observationsThisYear, 10 );
        } else {
          randomObservations = observationsThisYear;
        }

        // Get the top three of iconicTaxa observed over the year
        const reduced = observationsThisYear.reduce(
          ( iconicTaxa, observation ) => {
            const iconicTaxonId = observation.taxon.iconicTaxonId;
            iconicTaxa[iconicTaxonId] = iconicTaxa[iconicTaxonId] + 1 || 1;
            return iconicTaxa;
          },
          {}
        );
        const topThreeIconicTaxonIds = Object.keys( reduced )
          .sort( ( a, b ) => reduced[b] - reduced[a] )
          .slice( 0, 3 );

        const topThreeSpeciesBadges = [];
        const badges = realm.objects( "BadgeRealm" );
        topThreeIconicTaxonIds.forEach( ( id ) => {
          if ( id === null ) {
            return;
          }
          const highestEarned = badges
            .filtered( `iconicTaxonName != null AND iconicTaxonId == ${id}` )
            .sorted( "index", true )
            .sorted( "earned", true );
          const highestEarnedBadge = highestEarned[0];
          // Add the number of observations for this year to the badge info
          highestEarnedBadge.observationsThisYear = reduced[id];
          topThreeSpeciesBadges.push( highestEarnedBadge );
        } );

        const levelsEarned = badges
          .filtered( "iconicTaxonName == null AND earned == true" )
          .sorted( "count", true );

        // Get histogram data
        const histogram = observationsThisYear.reduce( ( data, observation ) => {
          const month = observation.date.getMonth();
          data[month] = { count: data[month]?.count + 1 || 1, month: month + 1  };
          return data;
        }, [] );
        for ( let i = 0; i < 12; i++ ) {
          if ( histogram[i] === undefined ) {
            histogram[i] = { count: 0, month: i + 1 };
          }
        }

        setState( {
          level: levelsEarned[0],
          observationsThisYear,
          topThreeSpeciesBadges,
          randomObservations,
          histogram,
          speciesCount
        } );
      } catch ( e ) {
        console.log( e, "couldn't open realm for fetching year in review stats" );
      }
    };
    fetchStatsFromRealm();
  }, [year] );

  return state;
};

const useFetchChallengesForYear = ( year ): any => {
  const [challengeBadges, setChallengeBadges] = useState( [] );
  const [challengeCount, setChallengeCount] = useState( undefined );

  useEffect( () => {
    const createBadge = ( latestBadge, numOfMonths ) => ( {
      name: "",
      availableDate: addMonths( latestBadge.availableDate, numOfMonths ),
      index: latestBadge.index + numOfMonths
    } );

    const createPlaceholderBadges = ( badges ) => {
      const remainderOfBadges = badges.length % 5;

      if ( remainderOfBadges === 0 || remainderOfBadges === 3 ) {
        // no placeholders needed
        return badges;
      }

      const badgePlaceholders = badges;
      const latestBadge = badges[badges.length - 1];

      let nextBadge = createBadge( latestBadge, 1 );

      // next challenge after June 2021 will be released Aug 2021
      if ( isEqual( new Date( 2021, 5, 1 ), new Date( latestBadge.availableDate ) ) ) {
        nextBadge = createBadge( latestBadge, 2 );
      }

      const badgeAfterNext = createBadge( nextBadge, 1 );

      if ( remainderOfBadges === 2 || remainderOfBadges === 4 ) {
        badgePlaceholders.push( nextBadge );
      }

      if ( remainderOfBadges === 1 ) {
        badgePlaceholders.push( nextBadge, badgeAfterNext );
      }

      return badgePlaceholders;
    };

    const fetchChallenges = async () => {
      try {
        // TODO: refactor into useCallbacks
        const firstOfYear = () => new Date( year, 0, 1, 0, 0, 0, 0 );
        const lastOfYear = () => new Date( year, 11, 31, 23, 59, 59 );

        const realm = await Realm.open( realmConfig );
        const challenges = realm
          .objects( "ChallengeRealm" )
          .filtered(
            "completedDate >= $0 && completedDate < $1",
            firstOfYear( year ),
            lastOfYear( year )
          )
          .sorted( "availableDate", false );
        const badges = challenges.map( ( challenge ) => challenge );
        setChallengeCount( badges.length );
        const badgesWithPlaceholders = createPlaceholderBadges( badges );
        setChallengeBadges( badgesWithPlaceholders );
      } catch ( e ) {
        console.log( e, "couldn't open realm: for fetching challenges for year in review" );
      }
    };
    fetchChallenges();
  }, [year] );

  return {challengeBadges, challengeCount};
};

export {
  useFetchStats,
  useCountObservationsForYear,
  useObservationsForYear,
  useFetchChallengesForYear
};
