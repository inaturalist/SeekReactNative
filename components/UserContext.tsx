import * as React from "react";

const UserContext: Object = React.createContext( null );
const ObservationContext: Object = React.createContext( null );
const SpeciesNearbyContext: Object = React.createContext( null );
const ChallengeContext: Object = React.createContext( null );
const SpeciesDetailContext: Object = React.createContext( null );

export {
  UserContext,
  ObservationContext,
  SpeciesNearbyContext,
  ChallengeContext,
  SpeciesDetailContext
};
