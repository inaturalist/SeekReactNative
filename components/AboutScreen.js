// @flow

import React, { useContext } from "react";
import {
  Image,
  View,
  TouchableOpacity
} from "react-native";
import { getVersion, getBuildNumber } from "react-native-device-info";
import { useNavigation } from "@react-navigation/native";
import type { Node } from "react";

import { viewStyles, imageStyles, textStyles } from "../styles/about";
import logos from "../assets/logos";
import i18n from "../i18n";
import { AppOrientationContext, UserContext } from "./UserContext";
import ScrollWithHeader from "./UIComponents/Screens/ScrollWithHeader";
import PrivacyAndTerms from "./UIComponents/PrivacyAndTerms";
import StyledText from "./UIComponents/StyledText";

const AboutScreen = (): Node => {
  const navigation = useNavigation();
  const appVersion = getVersion();
  const buildVersion = getBuildNumber();
  const { login } = useContext( UserContext );
  const { isTablet } = useContext( AppOrientationContext );

  const navToDebug = () => navigation.navigate( "DebugEmailScreen" );
  const disabled = !login;

  return (
    <ScrollWithHeader header="about.header" footer>
      <View style={[viewStyles.textContainer, isTablet && viewStyles.tabletContainer]}>
        <Image source={logos.iNat} />
        <View style={viewStyles.marginSmall} />
        <StyledText style={[textStyles.text, textStyles.boldText]}>{i18n.t( "about.seek" )}</StyledText>
        <StyledText style={textStyles.text}>{i18n.t( "about.joint_initiative" )}</StyledText>
        <View style={viewStyles.block} />
        <Image source={logos.casNatGeo} style={imageStyles.image} />
        <View style={viewStyles.marginLarge} />
        <StyledText style={textStyles.text}>{i18n.t( "about.original" )}</StyledText>
        <View style={viewStyles.margin} />
        <Image source={logos.wwfop} style={imageStyles.wwfop} />
        <View style={viewStyles.marginSmall} />
        <Image source={logos.hhmi} />
        <View style={viewStyles.margin} />
        <StyledText style={[textStyles.text, textStyles.boldText]}>{i18n.t( "about.designed_by" )}</StyledText>
        <StyledText style={textStyles.text}>{i18n.t( "about.inat_team_credits_3" )}</StyledText>
        <View style={viewStyles.marginSmallest} />
        <StyledText style={textStyles.text}>{i18n.t( "about.translators" )}</StyledText>
        <View style={viewStyles.marginSmallest} />
        <StyledText style={textStyles.text}>{i18n.t( "about.join_crowdin" )}</StyledText>
        <PrivacyAndTerms login={login} />
        <TouchableOpacity
          onPress={navToDebug}
          style={viewStyles.debug}
          disabled={disabled}
          testID="debug"
        >
          <StyledText style={textStyles.greenText}>
            {i18n.t( "about.version" ).toLocaleUpperCase()}
            {` ${appVersion} (${buildVersion})`}
          </StyledText>
        </TouchableOpacity>
        <StyledText style={textStyles.text}>
          {i18n.t( "about.help" )}
        </StyledText>
        <View style={viewStyles.block} />
      </View>
    </ScrollWithHeader>
  );
};

export default AboutScreen;
