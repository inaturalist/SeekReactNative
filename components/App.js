// @flow

import React, { useState, useEffect } from "react";
import * as RNLocalize from "react-native-localize";
import {
  setJSExceptionHandler,
  setNativeExceptionHandler
} from "react-native-exception-handler";
import type { Node } from "react";

import RootStack from "./Navigation/RootStack";
import { handleLocalizationChange, loadUserLanguagePreference } from "../utility/languageHelpers";
import { hideLogs } from "../utility/helpers";
import { LanguageContext } from "./UserContext";
import { getLanguage } from "../utility/settingsHelpers";
import SpeciesNearbyProvider from "./Providers/SpeciesNearbyProvider";
import UserLoginProvider from "./Providers/UserLoginProvider";
import ObservationProvider from "./Providers/ObservationProvider";
import AppOrientationProvider from "./Providers/AppOrientationProvider";
import ChallengeProvider from "./Providers/ChallengeProvider";
import SpeciesDetailProvider from "./Providers/SpeciesDetailProvider";
import { log } from "../react-native-logs.config";

const logger = log.extend( "App.js" );

const jsErrorHandler = ( e, isFatal ) => {
  // not 100% sure why jsErrorHandler logs e.name and e.message as undefined sometimes,
  // but I believe it relates to this issue, which reports an unnecessary console.error
  // under the hood: https://github.com/a7ul/react-native-exception-handler/issues/143

  // possibly also related to error boundaries in React 16+:
  // https://github.com/a7ul/react-native-exception-handler/issues/60
  if ( !e.name && !e.message ) {return;}
  logger.error( `JS Error: ${isFatal ? "Fatal:" : ""} ${e.stack}` );
};

// record JS exceptions
setJSExceptionHandler( jsErrorHandler );

// record native exceptions
// only works in bundled mode; will show red screen in dev mode
// tested this by raising an exception in RNGestureHandler.m
// https://stackoverflow.com/questions/63270492/how-to-raise-native-error-in-react-native-app
setNativeExceptionHandler( exceptionString => {
  logger.error( `Native Error: ${exceptionString}` );
} );

const App = ( ): Node => {
  useEffect( () => {
    logger.info( "App start" );
  }, [] );

  const [preferredLanguage, setLanguage] = useState( null );

  const getLanguagePreference = async ( ) => setLanguage( await getLanguage( ) );
  const toggleLanguagePreference = ( ) => getLanguagePreference( );

  const languageValue = { preferredLanguage, toggleLanguagePreference };

  useEffect( ( ) => {
    // wait until check for stored language is completed
    if ( !preferredLanguage ) { return; }
    loadUserLanguagePreference( preferredLanguage );
  }, [preferredLanguage] );

  useEffect( ( ) => {
    hideLogs( );

    // Context
    getLanguagePreference( );

    RNLocalize.addEventListener( "change", handleLocalizationChange );

    return ( ) => RNLocalize.removeEventListener( "change", handleLocalizationChange );
  }, [] );

  return (
    <AppOrientationProvider>
      <UserLoginProvider>
        <LanguageContext.Provider value={languageValue}>
          <ObservationProvider>
            <SpeciesNearbyProvider>
              <ChallengeProvider>
                <SpeciesDetailProvider>
                  <RootStack />
                </SpeciesDetailProvider>
              </ChallengeProvider>
            </SpeciesNearbyProvider>
          </ObservationProvider>
        </LanguageContext.Provider>
      </UserLoginProvider>
    </AppOrientationProvider>
  );
};

export default App;
