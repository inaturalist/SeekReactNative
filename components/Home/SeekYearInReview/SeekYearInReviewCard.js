// @flow

import React from "react";
import { View, Text } from "react-native";
import { useNavigation } from "@react-navigation/native";
import type { Node } from "react";

import i18n from "../../../i18n";
import {
  viewStyles,
  textStyles
} from "../../../styles/home/seekYearInReview";
import GreenButton from "../../UIComponents/Buttons/GreenButton";
import { AppOrientationContext, UserContext } from "../../UserContext";
import { useCountObservationsForYear } from "../../SeekYearInReview/hooks/seekYearInReviewHooks";


const SeekYearInReviewCard = ( ): Node => {
  // The year to show stats for
  const now = new Date();
  let year = now.getFullYear();
  const month = now.getMonth();
  // If it's January, show stats for the previous year
  if ( month === 0 ) {
    year -= 1;
  }
  const { navigate } = useNavigation();

  const { isLandscape } = React.useContext( AppOrientationContext );
  const { userProfile } = React.useContext( UserContext );
  const countObservationsThisYear = useCountObservationsForYear( year );

  const navToSeekYearInReview = () => navigate( "SeekYearInReview" );

  const isTime = month === 0 || month === 11;
  const hasObservations =
    !!countObservationsThisYear && countObservationsThisYear > 0;
  const showCard = userProfile?.isAdmin || ( isTime && hasObservations );

  if ( !showCard ) {
    return null;
  }

  return (
    <View testID="yir-card" style={viewStyles.whiteContainer}>
      <Text style={textStyles.header}>
        {i18n.t( "seek_year_in_review.header" ).toLocaleUpperCase()}
      </Text>
      <View style={viewStyles.textContainer}>
        <Text
          style={[
            textStyles.text,
            isLandscape && viewStyles.landscapeContainerRestrictedWidth
          ]}
        >
          {i18n.t( "seek_year_in_review.description" )}
        </Text>
      </View>
      <View style={viewStyles.marginGreenButtonLarge} />
      <GreenButton
        text="seek_year_in_review.button"
        handlePress={navToSeekYearInReview}
      />
      <View style={viewStyles.marginBottom} />
    </View>
  );
};

export default SeekYearInReviewCard;
