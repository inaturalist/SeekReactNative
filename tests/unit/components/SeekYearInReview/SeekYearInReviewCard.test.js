import React from "react";
import { render, screen } from "@testing-library/react-native";

import SeekYearInReviewCard from "../../../../components/Home/SeekYearInReview/SeekYearInReviewCard";
import { AppOrientationContext, UserContext } from "../../../../components/UserContext";
import { useCountObservationsForYear } from "../../../../components/SeekYearInReview/hooks/seekYearInReviewHooks";

jest.mock(
  "../../../../components/SeekYearInReview/hooks/seekYearInReviewHooks",
  () => ( {
    __esModule: true,
    useCountObservationsForYear: jest.fn( () => 100 )
  } )
);

const renderCard = () => {
  render(
    <AppOrientationContext.Provider value={{}} >
      <UserContext.Provider value={{}} >
        <SeekYearInReviewCard />
      </UserContext.Provider>
    </AppOrientationContext.Provider>
  );
};

const cardID = "yir-card";
describe( "SeekYearInReviewCard", () => {
    test( "should not render for zero observations this year", async () => {
        useCountObservationsForYear.mockReturnValueOnce( 0 );
        renderCard();
        expect( screen.queryByTestId( cardID ) ).toBeNull();
    } );

    test( "should not render outside December and January", async () => {
      jest.useFakeTimers().setSystemTime( new Date( 2020, 5, 15 ) );
      renderCard();
      jest.runOnlyPendingTimers();
      jest.useRealTimers();
      expect( screen.queryByTestId( cardID ) ).toBeNull();
    } );

    test( "should render in December with observations", async () => {
      jest.useFakeTimers().setSystemTime( new Date( 2020, 11, 15 ) );
      renderCard();
      jest.runOnlyPendingTimers();
      jest.useRealTimers();
      await screen.findByTestId( cardID );
      expect( screen ).toMatchSnapshot();
    } );

    test( "should render in January with observations", async () => {
      jest.useFakeTimers().setSystemTime( new Date( 2020, 0, 15 ) );
      renderCard();
      jest.runOnlyPendingTimers();
      jest.useRealTimers();
      await screen.findByTestId( cardID );
      expect( screen ).toMatchSnapshot();
    } );
} );
