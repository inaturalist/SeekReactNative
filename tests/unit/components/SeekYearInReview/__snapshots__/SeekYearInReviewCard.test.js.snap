// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeekYearInReviewCard should render in December with observations 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "paddingTop": 35,
    }
  }
  testID="yir-card"
>
  <Text
    style={
      {
        "color": "#38976d",
        "fontFamily": "Whitney-Semibold",
        "fontSize": 19,
        "letterSpacing": 1.12,
        "paddingLeft": 22,
      }
    }
  >
    SEEK YEAR IN REVIEW
  </Text>
  <View
    style={
      {
        "paddingHorizontal": 33,
        "paddingTop": 21,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#000000",
            "fontFamily": "Whitney-Book",
            "fontSize": 16,
            "lineHeight": 21,
          },
          undefined,
        ]
      }
    >
      Thanks for using Seek this year! Check out your personalized Year in Review here:
    </Text>
  </View>
  <View
    style={
      {
        "marginTop": 33,
      }
    }
  />
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignSelf": "center",
        "backgroundColor": "#38976d",
        "borderRadius": 34,
        "height": 79,
        "justifyContent": "center",
        "maxWidth": 317,
        "opacity": 1,
        "width": 317,
      }
    }
    testID="greenButton"
  >
    <Text
      allowFontScaling={true}
      style={
        [
          {
            "textAlign": "left",
          },
          [
            {
              "color": "#ffffff",
              "fontFamily": "Whitney-Semibold",
              "fontSize": 18,
              "letterSpacing": 1,
              "paddingTop": 3,
              "textAlign": "center",
            },
            {
              "letterSpacing": 1,
            },
            {
              "fontSize": 18,
            },
          ],
        ]
      }
    >
      VIEW MY SEEK REVIEW
    </Text>
  </View>
  <View
    style={
      {
        "marginTop": 48,
      }
    }
  />
</View>
`;

exports[`SeekYearInReviewCard should render in January with observations 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "paddingTop": 35,
    }
  }
  testID="yir-card"
>
  <Text
    style={
      {
        "color": "#38976d",
        "fontFamily": "Whitney-Semibold",
        "fontSize": 19,
        "letterSpacing": 1.12,
        "paddingLeft": 22,
      }
    }
  >
    SEEK YEAR IN REVIEW
  </Text>
  <View
    style={
      {
        "paddingHorizontal": 33,
        "paddingTop": 21,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#000000",
            "fontFamily": "Whitney-Book",
            "fontSize": 16,
            "lineHeight": 21,
          },
          undefined,
        ]
      }
    >
      Thanks for using Seek this year! Check out your personalized Year in Review here:
    </Text>
  </View>
  <View
    style={
      {
        "marginTop": 33,
      }
    }
  />
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignSelf": "center",
        "backgroundColor": "#38976d",
        "borderRadius": 34,
        "height": 79,
        "justifyContent": "center",
        "maxWidth": 317,
        "opacity": 1,
        "width": 317,
      }
    }
    testID="greenButton"
  >
    <Text
      allowFontScaling={true}
      style={
        [
          {
            "textAlign": "left",
          },
          [
            {
              "color": "#ffffff",
              "fontFamily": "Whitney-Semibold",
              "fontSize": 18,
              "letterSpacing": 1,
              "paddingTop": 3,
              "textAlign": "center",
            },
            {
              "letterSpacing": 1,
            },
            {
              "fontSize": 18,
            },
          ],
        ]
      }
    >
      VIEW MY SEEK REVIEW
    </Text>
  </View>
  <View
    style={
      {
        "marginTop": 48,
      }
    }
  />
</View>
`;
